contracts/contracts/harvest/BaseHarvester.sol

```diff
@@ -12,37 +12,76 @@ import { IVault } from "../interfaces/IVault.sol";
+import { IUniswapV3Router } from "../interfaces/uniswap/IUniswapV3Router.sol";
+import { IBalancerVault } from "../interfaces/balancer/IBalancerVault.sol";
+import { ICurvePool } from "../strategies/ICurvePool.sol";
 import "../utils/Helpers.sol";
 
 abstract contract BaseHarvester is Governable {
     using SafeERC20 for IERC20;
     using SafeMath for uint256;
     using StableMath for uint256;
 
-    event UniswapUpdated(address _address);
-    event SupportedStrategyUpdate(address _address, bool _isSupported);
+    enum SwapPlatform {
+        UniswapV2Compatible,
+        UniswapV3,
+        Balancer,
+        Curve
+    }
+
+    event SupportedStrategyUpdate(address strategyAddress, bool isSupported);
     event RewardTokenConfigUpdated(
-        address _tokenAddress,
-        uint16 _allowedSlippageBps,
-        uint16 _harvestRewardBps,
-        address _uniswapV2CompatibleAddr,
-        uint256 _liquidationLimit,
-        bool _doSwapRewardToken
+        address tokenAddress,
+        uint16 allowedSlippageBps,
+        uint16 harvestRewardBps,
+        SwapPlatform swapPlatform,
+        address swapPlatformAddr,
+        bytes swapData,
+        uint256 liquidationLimit,
+        bool doSwapRewardToken
+    );
+    event RewardTokenSwapped(
+        address indexed rewardToken,
+        address indexed swappedInto,
+        SwapPlatform swapPlatform,
+        uint256 amountIn,
+        uint256 amountOut
+    );
+    event RewardProceedsTransferred(
+        address indexed token,
+        address farmer,
+        uint256 protcolYield,
+        uint256 farmerFee
     );
+    event RewardProceedsAddressChanged(address newProceedsAddress);
+
+    error EmptyAddress();
+    error InvalidSlippageBps();
+    error InvalidHarvestRewardBps();
+
+    error InvalidSwapPlatform(SwapPlatform swapPlatform);
+
+    error InvalidUniswapV2PathLength();
+    error InvalidTokenInSwapPath(address token);
+    error EmptyBalancerPoolId();
+    error InvalidCurvePoolAssetIndex(address token);
+
+    error UnsupportedStrategy(address strategyAddress);
 
     // Configuration properties for harvesting logic of reward tokens
     struct RewardTokenConfig {
         // Max allowed slippage when swapping reward token for a stablecoin denominated in basis points.
         uint16 allowedSlippageBps;
         // Reward when calling a harvest function denominated in basis points.
         uint16 harvestRewardBps;
-        /* Address of Uniswap V2 compatible exchange (Uniswap V2, SushiSwap).
-         */
-        address uniswapV2CompatibleAddr;
+        // Address of compatible exchange protocol (Uniswap V2/V3, SushiSwap, Balancer and Curve).
+        address swapPlatformAddr;
         /* When true the reward token is being swapped. In a need of (temporarily) disabling the swapping of
          * a reward token this needs to be set to false.
          */
         bool doSwapRewardToken;
+        // Platform to use for Swapping
+        SwapPlatform swapPlatform;
```

Comment:
👍 Looks like this should upgrade okay, since it won't be clobbering anything, even though added into the middle of the struct.
<img width="1206" alt="image" src="https://github.com/OriginProtocol/origin-dollar/assets/837/0c96d170-4b90-47bf-908d-ac89ebdfe2f8">